src/redis-cli.c

```diff
@@ -418,91 +425,226 @@ static sds cliVersion(void) {
+                char *flag = flags->element[j]->str;
+                if (!strcmp(flag, "optional")) {
+                    optional = 1;
+                }
+                else if (!strcmp(flag, "multiple")) {
+                    multiple = 1;
+                }
+                else if (!strcmp(flag, "multiple_token")) {
+                    multipleToken = 1;
+                }
+            }
+        }
+    }
+
+    // Then build the "repeating part" of the argument string.
+    if (!strcmp(type, "key") ||
+        !strcmp(type, "string") ||
+        !strcmp(type, "integer") ||
+        !strcmp(type, "double") ||
+        !strcmp(type, "pattern") ||
+        !strcmp(type, "unix-time") ||
+        !strcmp(type, "token")) {
+            repeatPart = sdscat_orempty(repeatPart, name);
+    }
+    else if (!strcmp(type, "oneof")) {
+        repeatPart = cliConcatArguments(repeatPart, arguments, "|");
+    }
+    else if (!strcmp(type, "block")) {
+        repeatPart = cliConcatArguments(repeatPart, arguments, " ");
+    }
+    else if (strcmp(type, "pure-token") != 0) {
+        fprintf(stderr, "Unknown type '%s' set for argument '%s'\n", type, name);
+    }
+
+    // Finally, build the parameter string.
+    if (tokenPart[0] != '\0' && strcmp(type, "pure-token") != 0) {
+        tokenPart = sdscat(tokenPart, " ");
+    }
+    if (optional) {
+        params = sdscat(params, "[");
+    }
+    params = sdscat(params, tokenPart);
+    params = sdscat(params, repeatPart);
+    if (multiple) {
+        params = sdscat(params, " [");
+        if (multipleToken) {
+            params = sdscat(params, tokenPart);
+        }
+        params = sdscat(params, repeatPart);
+        params = sdscat(params, " ...]");
+    }
+    if (optional) {
+        params = sdscat(params, "]");
+    }
+    sdsfree(tokenPart);
+    sdsfree(repeatPart);
+    return params;
+}
+
+/* Initialize a command help entry for the command/subcommand described in reply.
+ * Returns the index of the next available position in the help entries table. */
+static int cliInitCommandHelpEntry(char *cmdname, char *subcommandname, int helpIndex, redisReply *reply) {
+    helpEntry *help = &(helpEntries[helpIndex++]);
+    help->argc = subcommandname ? 2 : 1;
+    help->argv = zmalloc(sizeof(sds) * help->argc);
+    help->argv[0] = sdsnew(cmdname);
+    sdstoupper(help->argv[0]);
+    if (subcommandname) {
+        help->argv[1] = sdsnew(subcommandname);
+        sdstoupper(help->argv[1]);
+    }
+    sds fullname = sdsnew(help->argv[0]);
+    if (subcommandname) {
+        fullname = sdscat(fullname, " ");
+        fullname = sdscat(fullname, help->argv[1]);
+    }
+    help->full = fullname;
+    help->type = CLI_HELP_COMMAND;
+
+    struct commandHelp *ch = zmalloc(sizeof(*ch));
+    help->org = ch;
+    ch->name = help->full;
+    ch->params = sdsempty();
+
+    if (reply->elements > 7) {
+        redisReply *map = reply->element[7];
```

Comment:
note to self, test how this works with old redis versions.

Reply:
No longer relevant - the code now only retrieves COMMAND DOCS, which doesn't exist on old versions.

Reply:
still need to test how it works on old versions (should at the very lest not explode).
is there absolutely no benefit of using the old COMMAND command?

Reply:
Now that you mention it... The old (i.e., current) implementation uses the output of COMMAND to generate stub help entries for commands not in the help.h file, with parameters named "key" and "arg". I presume this was used for dynamically-loaded commands such as modules. 

If we want, we can use this as a fallback mechanism in case COMMAND DOCS is unavailable, presumably when connecting to a previous version of the server. If we want to do that, we could call COMMAND INFO if COMMAND DOCS doesn't exist.

Reply:
the value of doing that seems low (showing "arg" and "key"), so we can choose to do that only if the effort is low too.
but what about showing a list of commands or auto-complete / suggest command names?

Reply:
COMMAND DOCS contains all the command and subcommand names, so that should already be covered.

Reply:
i meant that this part of the CLI help can keep working on old redis versions if we fallback to COMMAND command.

Reply:
Oh, right. Yes, that's an option. Ultimately, this is a product decision: what level of help support do we want to maintain for old server versions? And how much code is worth maintaining for that purpose?

Reply:
as i said, i don't see much value in the "arg" and "key" tips.
listing the commands is nice, but that's also low value, so only if the cost is low.

Reply:
Okay, I added the command names from COMMAND as a fallback. I don't think it's too much code to support, but it also gives only limited benefit. It doesn't even have subcommand names.